hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapRemoveAllMessageTask.java

```diff
@@ -16,34 +16,119 @@
 
 package com.hazelcast.client.impl.protocol.task.map;
 
+import com.hazelcast.client.impl.operations.OperationFactoryWrapper;
 import com.hazelcast.client.impl.protocol.ClientMessage;
 import com.hazelcast.client.impl.protocol.codec.MapRemoveAllCodec;
+import com.hazelcast.core.ExecutionCallback;
+import com.hazelcast.core.ICompletableFuture;
+import com.hazelcast.core.PartitioningStrategy;
 import com.hazelcast.instance.Node;
+import com.hazelcast.map.impl.MapContainer;
+import com.hazelcast.map.impl.MapService;
+import com.hazelcast.map.impl.MapServiceContext;
 import com.hazelcast.map.impl.operation.MapOperationProvider;
 import com.hazelcast.nio.Connection;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.query.PartitionPredicate;
 import com.hazelcast.query.Predicate;
 import com.hazelcast.security.permission.ActionConstants;
 import com.hazelcast.security.permission.MapPermission;
+import com.hazelcast.spi.Operation;
 import com.hazelcast.spi.OperationFactory;
+import com.hazelcast.spi.impl.operationservice.InternalOperationService;
+import com.hazelcast.spi.impl.operationservice.impl.operations.PartitionAwareOperationFactory;
 
 import java.security.Permission;
+import java.util.Collections;
 import java.util.Map;
 
 import static com.hazelcast.map.impl.EntryRemovingProcessor.ENTRY_REMOVING_PROCESSOR;
 import static com.hazelcast.map.impl.MapService.SERVICE_NAME;
+import static java.util.Collections.singletonList;
 
 public class MapRemoveAllMessageTask extends AbstractMapAllPartitionsMessageTask<MapRemoveAllCodec.RequestParameters> {
 
+    private Predicate predicate;
+
     public MapRemoveAllMessageTask(ClientMessage clientMessage, Node node, Connection connection) {
         super(clientMessage, node, connection);
     }
 
+    @Override
+    protected void processMessage() {
+        if (!(predicate instanceof PartitionPredicate)) {
```

Comment:
Should this not be taken of on the client proxy so that the client will immediately send it to the right member?

Reply:
Yes, that's handled by https://github.com/hazelcast/hazelcast/pull/14100/files#diff-3540ff2664ebe82892440b80a7293680R388, but to maintain the backward compatibility with the older clients we have to support non-partition and partition predicates here, in the later case partition ID won't be set in the message itself if it was sent by an old client.